### ISSUE TITLE
Improve accessibility for screen reader users

### ISSUE THREAD

**Issue by Neurrone** [jeffvli/sonixd#234]
I'm visually impaired and use a piece of software called a screen reader to use the computer.

I found that for the most part, I am able to use Sonixd. Being written in Electron, it is already mostly accessible since it is normal HTML.

**Describe the bug**

There are various elements of the UI that don't read properly with a screen reader:

1. Buttons with purely iconic labels
2. Buttons with a label that is exposed via a tooltip. Due to how rsuite doesn't implement the aria tooltip pattern, those labels don't read for me as well right now
3. Divs with iconic labels and onclick handlers that are actually buttons. E.g, the play / seek back / seek forward row of buttons
4. The seek and volume progress bars (due to rsuite not implementing accessibility for it in v4, it looks like v5 fixes this somewhat though)

**To Reproduce**

1. Install the free [NVDA screen reader on Windows](https://www.nvaccess.org/).
2. Try using it with the app
3. Notice that numerous elements are read as "button", "clickable" which don't convey any information about its purpose. Its the equivalent of all buttons using the same visual icon with no label.

**Expected behavior**

Controls read out properly. E.g, "play", "add to queue" etc.

**Desktop (please complete the following information):**
 - OS: windows 10
 - Version: master

**Comment by Neurrone** [comment]
I've raised #235 which fixes the first 3 items as these are the lowest hanging fruit.

**Comment by jeffvli** [comment]
I'm not opposed to switching to a different slider component if you know of any that can better handle the screen reader.

**Comment by Neurrone** [comment]
The following would work:

* The built-in HTML5 slider element, this would be the best approach. I'm wondering if there were any constraints that prevented that element from being used?
* Augmenting the current markup by implementing the [slider pattern](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_slider_role). This role and the associated attributes is a way of notifying the screen reader that the elements has slider semantics, the min and max values etc. I would typically only reserve this for cases where e.g, a div is being used and I couldn't switch to using the native html element.

**Comment by jeffvli** [comment]
I guess I just used rsuite's component to try to keep things as consistent as possible in regards to styling. Using a html5 slider would indeed probably be the best solution, as there aren't really any special constraints (besides needing to restyle it). Do you need to attach attributes to it or anything else for it to work properly with the reader?

**Comment by Neurrone** [comment]
Nope, no special attributes are needed. It natively understands HTML elements. The only one I may add is `aria-label` so that it says something like "seek progress bar" or "volume progress bar" if the label attached to it is "seek" or "volume".

I would also be open to raising a PR to rsuite because it would be ideal if this is fixed there. But I see that we're currently using v4 instead of v5,.

The role="..." attributes etc that you saw me add previously was to have it treat e.g, various elements as buttons instead of what the html said, which isn't ideal but lets us avoid changing too much markup.

**Comment by jeffvli** [comment]
I see, thanks.

At this point I'd say not to bother creating an issue with rsuite (unless you'd want to let them know separately). There were quite a few breaking changes from v4 to v5 which ended up in me scrapping the plan to upgrade, so Sonixd is probably going to be on v4 for the foreseeable future.

I also find that rsuite's slider component is a bit clunky, so creating a new one would probably be better.

**Comment by Neurrone** [comment]
I'm playing with this more and the main challenge I'm running to is using context menus. 

I'm trying to add a bunch of songs to a playlist. I use the computer exclusively via the keyboard.

I notice that you can press the applications menu to get the context menu to appear. However, keyboard focus doesn't get moved to the first menu item when it appears.

The ideal end state would be:

* Press applications, keyboard focus moves to the first menu item
* The ability to navigate through it like a normal system context menu. I.e, pressing up / down arrow moves between items, pressing right arrow expands something with submenus. This would be nice to have but not as impactful as just having focus move to the menu automatically when its opened.
* When exiting the menu, focus gets restored back to the previous position

I'll see if its possible to implement this.

**Comment by Neurrone** [comment]
Where is the code which controls the initial display of the menu? I'm hoping to start by saving the element that currently has keyboard focus to be restored when the menu closes.

**Comment by jeffvli** [comment]
> Where is the code which controls the initial display of the menu? I'm hoping to start by saving the element that currently has keyboard focus to be restored when the menu closes.

Here's where the context menu is initially opened from the list view.

https://github.com/jeffvli/sonixd/blob/943101a1d2d2635dddb6fba40a29a24a9f1c022a/src/components/viewtypes/ListViewTable.tsx#L464-L518

**Comment by Neurrone** [comment]
I found a bigger issue, which is that its really hard for me to scroll through tables because they are virtualized.

When I use the keyboard to read each row with my screen reader, I find that reaching the bottom, it doesn't immediately load the next page of results. I also couldn't figure out how to scroll upwards.

Since this uses rsuite, I don't know if its possible to fix this because this is likely an issue with that library 🤔

**Comment by jeffvli** [comment]
I have a hook [useListScroll](https://github.com/jeffvli/sonixd/blob/main/src/hooks/useListScroll.ts) that you could probably assign to hotkeys to scroll the list up/down.

All you need to do is assign it to the ref of a `ListViewType` component and you can use it as `listScroll(height)` to scroll the list. You can see an example of the implementation in the [AlbumList](https://github.com/jeffvli/sonixd/blob/main/src/components/library/AlbumList.tsx).

I can take a look at this soon, as I have a couple ideas on how I want to change the keyboard navigation of the list view.

**Comment by Neurrone** [comment]
Since keyboard navigation in lists is being improved / updated, would you be opposed to introducing a new dependency to improve the experience for keyboard-only users?

The interaction that I have in mind is the ability to not only scroll the viewport, but to be able to get to individual cells via the arrow keys. I think this is important in this case because the tables do have controls that can be activated.

Its possible to code this out manually, but I've found that the [`useTable` hook from React-aria](https://react-spectrum.adobe.com/react-aria/useTable.html) already implements this.

**Comment by jeffvli** [comment]
The `useTable` hooks like interesting. I'll see if I can incorporate it my rsuite-table implementation.

**Comment by Neurrone** [comment]
Thanks. Let me know if you run into any issues with doing that and I'll see how I can help.
